### PR TITLE
This Week in Glean: Bytes in Memory (on Android)

### PR DESCRIPTION
**Pull request by badboy**
[rendered](https://github.com/badboy/fnordig.de/blob/new-twig/_posts/2020-05-04-this-week-in-glean.md)

cc @mdboom @Dexterp37 @travis79 @chutten

### CODE REVIEW

File: _posts/2020-05-04-this-week-in-glean.md [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+(“This Week in Glean” is a series of blog posts that the Glean Team at Mozilla is using to try to communicate better about our work. They could be release notes, documentation, hopes, dreams, or whatever: so long as it is inspired by Glean.)
+
+Last week's blog post: [ This Week in Glean: Glean for Python on Windows](https://blog.mozilla.org/data/2020/04/27/this-week-in-glean-glean-for-python-on-windows/)
+by Mike Droettboom.
+All "This Week in Glean" blog posts are available on the [Mozilla Data blog](https://blog.mozilla.org/data/category/glean/).
+
+---
+
+With the Glean SDK we follow in the footsteps of [other teams](https://github.com/mozilla/application-services) to build a cross-platform library to be used in mobile applications.
```
Comment (Dexterp37) [inline _posts/2020-05-04-this-week-in-glean.md]: We're now beyond mobile, right? :-)

File: _posts/2020-05-04-this-week-in-glean.md [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+}
+
+struct Request {
+    id: String,
+    url: String,
+}
+```
+
+(This code is simplified for the sake of this blog post.
+You can find the full code online [in the Glean repository](https://github.com/mozilla/glean/blob/8e0599b4df6c1a08f985e4ad5328fcf81a56a084/glean-core/src/upload/mod.rs).)
```
Comment (Dexterp37) [inline _posts/2020-05-04-this-week-in-glean.md]:
```suggestion
You can find the full code online [in the Glean repository](https://github.com/mozilla/glean/blob/8e0599b4df6c1a08f985e4ad5328fcf81a56a084/glean-core/src/upload/mod.rs#L38-L48).)
```

File: _posts/2020-05-04-this-week-in-glean.md [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+(“This Week in Glean” is a series of blog posts that the Glean Team at Mozilla is using to try to communicate better about our work. They could be release notes, documentation, hopes, dreams, or whatever: so long as it is inspired by Glean.)
+
+Last week's blog post: [ This Week in Glean: Glean for Python on Windows](https://blog.mozilla.org/data/2020/04/27/this-week-in-glean-glean-for-python-on-windows/)
+by Mike Droettboom.
+All "This Week in Glean" blog posts are available on the [Mozilla Data blog](https://blog.mozilla.org/data/category/glean/).
+
+---
+
+With the Glean SDK we follow in the footsteps of [other teams](https://github.com/mozilla/application-services) to build a cross-platform library to be used in mobile applications.
+In this blog post we're going to take a look at how we transport some rich data across the FFI boundary to be reused on the Kotlin side of things.
```
Comment (Dexterp37) [inline _posts/2020-05-04-this-week-in-glean.md]: Can you provide a bit more context about what the article is about? By reading this I'd expect the article to talk about FFI stuff , which is something you mostly worked on. However, the article also describes bits of the upload mechanism, which is what most of the implementation is about. I'd spend a few more words on this. If this is about describing the upload mechanism as well, then I'd provide a super tl;dr about how it should work.

File: _posts/2020-05-04-this-week-in-glean.md [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+when (val action = incomingTask.toTask()) {
+    is Task.Upload -> upload(action.request.id, aciton.request.url)
+    Task.Wait -> return Result.retry()
+    Task.Done -> return Result.success()
+}
+```
+
+## What's next?
+
+Currently this piece of code is under testing.
```
Comment (Dexterp37) [inline _posts/2020-05-04-this-week-in-glean.md]: The upload parts are under testing, but the more general "passing data through FFI" part has been around for a long while. I'd recommend clarifying that a bit.

**Review by Dexterp37 — CHANGES_REQUESTED** [review]

File: _posts/2020-05-04-this-week-in-glean.md (hunk 3 above)
Comment (badboy) [reply _posts/2020-05-04-this-week-in-glean.md]: ok with linking [to the docs?](https://github.com/mozilla/glean/blob/8e0599b4df6c1a08f985e4ad5328fcf81a56a084/docs/dev/core/internal/upload.md)

this is really more about the approach than about uploading.

Comment (Dexterp37) [reply _posts/2020-05-04-this-week-in-glean.md]: > 
> 
> ok with linking [to the docs?](https://github.com/mozilla/glean/blob/8e0599b4df6c1a08f985e4ad5328fcf81a56a084/docs/dev/core/internal/upload.md)
Yes !

> this is really more about the approach than about uploading.

Perfect! Even just static exactly this somewhere in this paragraph would make the rest more coherent, I believe.

**Review by Dexterp37 — APPROVED** [review]

**Review by mdboom — APPROVED** [review]
LGTM.  Are you going to publish on blog.mozilla.org/data as well as this blog?

**Comment by badboy** [comment]
> LGTM. Are you going to publish on blog.mozilla.org/data as well as this blog?

Yup.

File: _posts/2020-05-04-this-week-in-glean.md [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+  tags:
+    - mozilla
+    - rust
+---
+
+(“This Week in Glean” is a series of blog posts that the Glean Team at Mozilla is using to try to communicate better about our work. They could be release notes, documentation, hopes, dreams, or whatever: so long as it is inspired by Glean.)
+
+Last week's blog post: [ This Week in Glean: Glean for Python on Windows](https://blog.mozilla.org/data/2020/04/27/this-week-in-glean-glean-for-python-on-windows/)
+by Mike Droettboom.
+All "This Week in Glean" blog posts are available on the [Mozilla Data blog](https://blog.mozilla.org/data/category/glean/).
```
Comment (chutten) [inline _posts/2020-05-04-this-week-in-glean.md]: I thought that the index in the Book of Glean was going to be the canonical index?

File: _posts/2020-05-04-this-week-in-glean.md [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+
+> *Note:* This blog post is not a good introduction on doing Rust on Android, but I do plan to write about that in the future as well.
+
+Most of this implementation was done by [Bea](https://brizental.github.io/) and I have been merely a reviewer,
+around for questions and currently responsible for running final tests on this feature.
+
+## The problem
+
+The Glean SDK provides an [FFI API](https://github.com/mozilla/glean/tree/master/glean-core/ffi) that can be consumed by what we call language bindings.
+For the most part we pass <abbr title="Plain Old Data">POD</abbr> over that API boundary:
```
Comment (chutten) [inline _posts/2020-05-04-this-week-in-glean.md:35]: Do `abbr` render properly on fnordig? On github's preview they're eaten.

File: _posts/2020-05-04-this-week-in-glean.md [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+integers of various sizes (a problem in and of itself if the sides disagree about certain integer sizes and signedness),
+bools (but actually encoded as an 8-bit integer),
```
Comment (chutten) [inline _posts/2020-05-04-this-week-in-glean.md]: I'd love a couple of words here as to why.

File: _posts/2020-05-04-this-week-in-glean.md [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+but also strings as pointers to null-terminated UTF-8 strings in memory (sometimes we encode data into JSON and pass that over as strings).
```
Comment (chutten) [inline _posts/2020-05-04-this-week-in-glean.md]:
```suggestion
but also strings as pointers to null-terminated UTF-8 strings in memory (for test APIs we encode data into JSON and pass that over as strings).
```

File: _posts/2020-05-04-this-week-in-glean.md [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+    object Done : Task()
+}
+```
+
+As the final piece of this code on the Kotlin side we can now fetch new tasks, convert it to more convenient Kotlin objects and work with them:
+
+```java
+val incomingTask = LibGleanFFI.INSTANCE.glean_get_next_task()
+when (val action = incomingTask.toTask()) {
+    is Task.Upload -> upload(action.request.id, aciton.request.url)
```
Comment (chutten) [inline _posts/2020-05-04-this-week-in-glean.md]:
```suggestion
    is Task.Upload -> upload(action.request.id, action.request.url)
```

**Review by chutten — APPROVED** [review]
Awesome Sauce

File: _posts/2020-05-04-this-week-in-glean.md (hunk 5 above)
Comment (badboy) [reply _posts/2020-05-04-this-week-in-glean.md]: Will change that

File: _posts/2020-05-04-this-week-in-glean.md (hunk 6 above)
Comment (badboy) [reply _posts/2020-05-04-this-week-in-glean.md:35]: yes, they do!

**Review by travis79 — APPROVED** [review]
Great article @badboy!!

**Pull request merged**